packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol

```diff
@@ -0,0 +1,180 @@
+        require(address(destinationToken) != address(0));
+        uint256 sourceBalanceBefore = sourceToken.balanceOf(msg.sender);
+        uint256 sourceSupplyBefore = sourceToken.totalSupply();
+        uint256 destinationBalanceBefore = destinationToken.balanceOf(msg.sender);
+        uint256 destinationSupplyBefore = destinationToken.totalSupply();
+
+        vm.prank(msg.sender);
+        try sourceToken.sendERC20(msg.sender, amount, destinationChainId) {
+            uint256 sourceBalanceAfter = sourceToken.balanceOf(msg.sender);
+            uint256 destinationBalanceAfter = destinationToken.balanceOf(msg.sender);
+            // no free mint
+            assert(sourceBalanceBefore + destinationBalanceBefore == sourceBalanceAfter + destinationBalanceAfter);
+            // 22
+            assert(sourceBalanceBefore - amount == sourceBalanceAfter);
+            assert(destinationBalanceBefore + amount == destinationBalanceAfter);
+            uint256 sourceSupplyAfter = sourceToken.totalSupply();
+            uint256 destinationSupplyAfter = destinationToken.totalSupply();
+            // 23
+            assert(sourceSupplyBefore - amount == sourceSupplyAfter);
+            assert(destinationSupplyBefore + amount == destinationSupplyAfter);
+        } catch {
+            assert(address(destinationToken) == address(sourceToken) || sourceBalanceBefore < amount);
+        }
+    }
+
+    function fuzz_MintSupertoken(uint256 index, uint96 amount) external {
+        index = bound(index, 0, allSuperTokens.length - 1);
+        address addr = allSuperTokens[index];
+        vm.prank(BRIDGE);
+        // medusa calls with different senders by default
+        OptimismSuperchainERC20(addr).mint(msg.sender, amount);
+        // currentValue will be zero if key is not present
+        (,uint256 currentValue) = ghost_totalSupplyAcrossChains.tryGet(MESSENGER.superTokenInitDeploySalts(addr));
+        ghost_totalSupplyAcrossChains.set(MESSENGER.superTokenInitDeploySalts(addr), currentValue + amount);
+    }
+
+    // TODO: will need rework after
+    //   - non-atomic bridge
+    //   - `convert`
+    /// @custom:property-id 24
+    /// @custom:property sum of supertoken total supply across all chains is always equal to convert(legacy, super)-
+    /// convert(super, legacy)
+    /// @dev deliberately not a view method so medusa runs it but not the view methods defined by Test
+    function property_totalSupplyAcrossChainsEqualsMints() external {
+        for (uint256 i = 0; i < ghost_totalSupplyAcrossChains.length(); i++) {
+            uint256 totalSupply = 0;
+            (bytes32 currentSalt, uint256 trackedSupply) = ghost_totalSupplyAcrossChains.at(i);
+            for (uint256 j = 0; j < MAX_CHAINS; j++) {
+                address supertoken = MESSENGER.superTokenAddresses(j, currentSalt);
+                if (supertoken != address(0)) {
+                    totalSupply += OptimismSuperchainERC20(supertoken).totalSupply();
```

Comment:
Could you add some comments or use another var names instead of `_i` of `_j`? bc is kinda difficult to understand hehe